File: src/components/playlists/clocks.js
import { getClocks, getMusicForTag } from "../itframe/api"
import _ from "underscore"

export default async () => {
    debug("Loading clocks")
    const clocks = await getClocks()
    const currentClock = getClockForDayHourMinute(clocks, new Date().getDay(), new Date().getHours(), new Date().getMinutes())
    debug(currentClock)
    const songs = {}
    for (let tag of currentClock.tags) {
        songs[tag.tag] = await getMusicForTag(tag.tag)
    }
    const next100Songs = generatePlaylist(currentClock.tags, songs)
    debug("Clocks ready")
    return next100Songs
}


const generatePlaylist = function (tags, songsForTag) {
    const playlist = [] // max 100 songs

    for (var tag of tags) {
        debug("Selecting " + tag.percent + " songs from " + tag.tag)
        songsForTag[tag.tag] = _.shuffle(songsForTag[tag.tag])
        for (let i = 0; i < tag.percent; i++) { // select num procent songs
            if (songsForTag[tag.tag].hasOwnProperty(i)) {
                debug("Adding " + songsForTag[tag.tag][i]._id)
                playlist.push(songsForTag[tag.tag][i])
            } else {
                const song = songsForTag[tag.tag][Math.floor((Math.random() * (songsForTag[tag.tag].length - 1)))]
                if (song) {
                    debug("Adding random song " + song._id)
                    playlist.push(song)
                }
            }
        }
    }
    return _.shuffle(playlist)
}

const getClockForDayHourMinute = (clocks, day, hour, minute) => {
    for (let id in clocks) {
        if (clocks.hasOwnProperty(id)) {
            if (clocks[id].start.dayOfWeek < day && clocks[id].end.dayOfWeek > day) {
                // ] day [
                return clocks[id];
            } else if ((clocks[id].start.dayOfWeek === day || clocks[id].end.dayOfWeek === day)) {
                // [ day ]
                if ((clocks[id].start.dayOfWeek === day && clocks[id].start.hour <= hour) || (clocks[id].end.dayOfWeek === day && clocks[id].end.hour >= hour)) {
                    // check end minutes
                    // [ day ] [ hour ]
                    if (clocks[id].start.hour < hour && clocks[id].end.hour > hour) {
                        // ] hour [
                        return clocks[id];
                    } else if ((clocks[id].start.hour === hour && clocks[id].start.minute >= minute) || (clocks[id].end.hour === hour && clocks[id].end.minute >= minute)) {
                        // [ day ] [ hour ] [ minute ]
                        return clocks[id];
                    }
                }
            }
        }
    }
    return null;
}

File: playlists/clocks/clocks.go
package clocks

import (
	"math/rand"
	"strconv"
	"time"

	"github.com/innovate-technologies/DJ/cron"
	"github.com/innovate-technologies/DJ/data"
	"github.com/innovate-technologies/DJ/itframe"
)

type itframeAPI interface {
	GetAllClocks() []data.Clock
	GetAllSongsForTag(tag string) []data.Song
}

var api itframeAPI = itframe.New()

// GetSongs returns 100 songs for the current clock
func GetSongs() (songs []data.Song) {
	songs = []data.Song{}
	clock := getCurrentClock()

	for _, tag := range clock.Tags {
		songs = append(songs, getSongsForTag(tag.Tag, tag.Percent)...)
	}
	shuffle(songs)
	return
}

// SetReloads sets the reload events when a new clock has to start
func SetReloads() {
	allClocks := api.GetAllClocks()
	c := cron.GetInstance()

	for _, clock := range allClocks {
		c.Add(cron.Action{
			DayOfMonth: "*",
			Month:      "*",
			DayOfWeek:  strconv.FormatInt(int64(clock.Start.DayOfWeek), 10),
			Hour:       strconv.FormatInt(int64(clock.Start.Hour), 10),
			Minute:     strconv.FormatInt(int64(clock.Start.Minute), 10),
			Event:      "reloadQueue",
		})
	}
}

func getCurrentClock() data.Clock {
	now := time.Now()

	day := int(now.Weekday())
	if day == 0 {
		day = 7
	}

	allClocks := api.GetAllClocks()

	for _, clock := range allClocks {
		if clock.Start.DayOfWeek < day && clock.End.DayOfWeek > day {
			// ] day [
			return clock
		}
		if clock.Start.DayOfWeek == day || clock.End.DayOfWeek == day {
			// [ day ]
			if (clock.Start.DayOfWeek == day && clock.Start.Hour <= now.Hour()) || (clock.End.DayOfWeek == day && clock.End.Hour >= now.Hour()) {
				// check end minutes
				// [ day ] [ hour ]
				if clock.Start.Hour < now.Hour() && clock.End.Hour > now.Hour() {
					// ] hour [
					return clock
				} else if (clock.Start.Hour == now.Hour() && clock.Start.Minute >= now.Minute()) || (clock.End.Hour == now.Hour() && clock.End.Minute >= now.Minute()) {
					// [ day ] [ hour ] [ minute ]
					return clock
				}
			}
		}
	}

	return data.Clock{}
}

func getSongsForTag(tag string, num int) []data.Song {
	allSongs := api.GetAllSongsForTag(tag)
	shuffle(allSongs)
	if len(allSongs) >= num {
		return allSongs[:num]
	}

	for len(allSongs) < num {
		if len(allSongs) > 1 {
			allSongs = append(allSongs, allSongs[rand.Intn(len(allSongs)-1)])
		} else {
			allSongs = append(allSongs, allSongs[0])
		}
	}
	return allSongs
}

func shuffle(vals []data.Song) {
	r := rand.New(rand.NewSource(time.Now().Unix()))
	for len(vals) > 0 {
		n := len(vals)
		randIndex := r.Intn(n)
		vals[n-1], vals[randIndex] = vals[randIndex], vals[n-1]
		vals = vals[:n-1]
	}
}
